### ISSUE TITLE
[Plugin] Slow Handling Of Huge Data By Proxy Request

### ISSUE THREAD

**Issue by softhub-software-development** [abhinavsingh/proxy.py#980]
I notice that queueing a huge response in a plugins `"handle_client_request()"` method is very slow.

```
def handle_client_request(self, request: HttpParser) -> Optional[HttpParser]:
    if some condition:
        return handle_request_locally()
    # else access remote resource
    return request

def handle_request_locally(self) -> None:
    f = open('my_file', 'rb')
    file_data = f.read()
    f.close()
    self.client.queue(
        okResponse(
            file_data,
            {b'Content-Type': b'application/octet-stream'},
            conn_close=True,
    ))
```
Connecting to the proxy via a browser and thereby accessing a huge file results in very slow
handling of the data. My question: is there a bottleneck somewhere in proxy.py or is it simply the wrong
approach? Even when running proxy.py on localhost and no network transfer is involved the browser
indicates that it will take hours or days to transfer the data.

**Comment by abhinavsingh** [comment]
How does a speed test when `proxy.py` is enabled looks like for you?

1. No Proxy: ~250 Mbps
2. With Proxy: ~250 Mbps

^^^ Metrics I see at my end.

**Comment by abhinavsingh** [comment]
Simply visit [fast.com](https://fast.com)

**Comment by abhinavsingh** [comment]
Are you still facing issues with this?  Please re-open and we can dig into it if necessary.  Thank you.

**Comment by abhinavsingh** [comment]
One optimization you could do is to queue the packets in chunk i.e. by reading the file in chunk itself.  This should improve performance automagically, as chunk of files are streamed to the client as they are read from the disk.  Use chunked encoded response for responsive client experience.  By queueing a 2Gb file without chunked encoding, there is all of a sudden pressure on the system, trying to flush out that big 2Gb memory view.

Irrespective, if you still experience any issue, kindly reopen the ticket.  Also post your speed test results.  Speed test is basically a verification that there is no problem in `proxy.py` core.  With or without proxy, you should get same throughput via speed test results.

**Comment by softhub-software-development** [comment]
I noticed that increasing the DEFAULT_MAX_SEND_SIZE in proxy.common.constants.py from 16k to 128k improves performance significantly.

**Comment by abhinavsingh** [comment]
@softhub-software-development Thanks for finding this abnormality.  Certainly missed it.  I recently updated `DEFAULT_BUFFER_SIZE`.  Ideally, `DEFAULT_MAX_SEND_SIZE` should also be configurable just like client & server recv buffer sizes.

I have re-opened this ticket to track the same.  Thank you!!!

**Comment by abhinavsingh** [comment]
@achen353 Kindly check finding by @softhub-software-development , may be it can help you with https://github.com/abhinavsingh/proxy.py/issues/1044 too, can you try and let us know.  Thank you!!!

**Comment by softhub-software-development** [comment]
Setting the DEFAULT_MAX_SEND_SIZE to 128k results in an error message "OSError when flushing buffer to client". Increasing it from 16k to 64k seems ok.

**Comment by abhinavsingh** [comment]
> "OSError when flushing buffer to client". Increasing it from 16k to 64k seems ok.

Indeed, we cannot send more than what our kernel allows.  This is a setting controlled by `sysctl`, but I don't remember which one exactly :( , will try to look back into it, but just search up and something will show up.

- There might be other OS/Socket level settings which can impact this
- Also, do remember you can always use functionalities available on socket but not in `proxy.py` connection classes e.g. [sendfile](https://docs.python.org/3/library/socket.html#socket.socket.sendfile)

Let me know your findings.  Happy to move these defaults within core so that everyone can benefit from these.

**Comment by abhinavsingh** [comment]
https://github.com/abhinavsingh/proxy.py/pull/1060 will add a `--max-sendbuf-size` flag which you can use to customize this value.  Also, now this value defaults to 64 Kb as you experimented.  Will wait for CI and some local laptop runtime before merge.  Feel free to try out directly from PR branch.

**Comment by achen353** [comment]
> @achen353 Kindly check finding by @softhub-software-development , may be it can help you with #1044 too, can you try and let us know. Thank you!!!

@abhinavsingh Thank you for the advice. I've been attempting to get access to our team's public server for testing. I'll definitely try this out and keep you posted.

**Comment by softhub-software-development** [comment]
I can observe a 50% performance increase by using the data type "bytes" instead of "memoryview" in core.connection.py:


```
    def __init__(self, tag: int) -> None:
        ...
        self.buffer: List[bytes] = []
       ...

    def queue(self, mv: memoryview) -> None:
        self.buffer.append(mv.tobytes())
        self._num_buffer += 1

    def flush(self) -> int:
        ...
        mv = self.buffer[0]       #.tobytes()
        sent: int = self.send(mv[:DEFAULT_MAX_SEND_SIZE])
        if sent == len(mv):
            self.buffer.pop(0)
            self._num_buffer -= 1
        else:
            self.buffer[0] = mv[sent:]
        #del mv
        ...

```

mv is now a misnomer, of course.

**Comment by abhinavsingh** [comment]
@softhub-software-development I think real issue is not memory view but this

- https://github.com/abhinavsingh/proxy.py/blob/develop/proxy/core/connection/connection.py#L88-L90

```python
        # TODO: Assemble multiple packets if total
        # size remains below max send size.
        sent: int = self.send(mv[:max_send_size])
```

**Comment by abhinavsingh** [comment]
> Avoiding len(mv) would help to further optimize this routine.

+1 to this.  Indeed, I think such optimizations will benefit us greatly

**Comment by abhinavsingh** [comment]
> I can observe a 50% performance increase by using the data type "bytes" instead of "memoryview"

I think you mean, increase in performance after removing `.tobytes()`, correct.  I agree, we can benefit from this too.

**Comment by abhinavsingh** [comment]
@softhub-software-development Opened #1063 to track `connection` related optimizations.  Thanks a lot for sharing these findings with us :D
